### ISSUE TITLE
Paski 

### ISSUE THREAD

**Issue by KMoszczyc** [codziennatranskrypcjatvpis/codziennatranskrypcjatvpis.github.io#2]
Jest szansa na paski z wiadomości na dole ekranu czy raczej nie da się ich wyciągnąć? 

A i super pomysł, mega szacunek za trzymanie tego raspberry włączonego przez rok, dzień w dzień.

**Comment by codziennatranskrypcjatvpis** [comment]
> A i super pomysł, mega szacunek za trzymanie tego raspberry włączonego przez rok, dzień w dzień.

Dzięki, po to kupiłem raspberry, aby mieć usługi działające 24/7.
> Jest szansa na paski z wiadomości na dole ekranu czy raczej nie da się ich wyciągnąć?

Myślałem nad tym kiedyś i porzuciłem, ponieważ nie mam dostatecznej wiedzy z OpenCV oraz zbieraniem pasków zajmuje się [@FlasH](https://twitter.com/FlasH_vikop) już. Ale jeśli napiszesz coś, będzie to działało, wykonywało się w sensownym czasie na RPi - to będą i paski.

**Comment by KMoszczyc** [comment]
Hejka @codziennatranskrypcjatvpis ,
zrobiłem skrypt pythonowy, który zdaje się całkiem nieźle zbiera paski. (https://github.com/KMoszczyc/TVP-Headlines). OpenCV + Tesseract (biblioteka do OCR którą trzeba osobno zainstalować i podać ścieżkę w skrypcie, bo pytesseract to tylko wrapper tesseracta). Na ryzenie 3600 ciągnie kilka razy szybciej niż w czasie rzeczywistym, więc na RPi powinno pójść gładko. 

Póki co to jest zrobione pod ekstrakcje nagłówków z pliku mp4, ale podmiana tego na inne źródło nie powinna być trudna. Mogę to zrobić, tylko nie wiem jak w swoim setupie zbierasz obraz z reżimówki. :)

**Comment by StachowiakDawid** [comment]
> Hejka @codziennatranskrypcjatvpis , zrobiłem skrypt pythonowy, który zdaje się całkiem nieźle zbiera paski. (https://github.com/KMoszczyc/TVP-Headlines). OpenCV + Tesseract (biblioteka do OCR którą trzeba osobno zainstalować i podać ścieżkę w skrypcie, bo pytesseract to tylko wrapper tesseracta). Na ryzenie 3600 ciągnie kilka razy szybciej niż w czasie rzeczywistym, więc na RPi powinno pójść gładko.
> 
> Póki co to jest zrobione pod ekstrakcje nagłówków z pliku mp4, ale podmiana tego na inne źródło nie powinna być trudna. Mogę to zrobić, tylko nie wiem jak w swoim setupie zbierasz obraz z reżimówki. :)

Wybacz, że tak długo tutaj nie zaglądałem, ale postaram się jak najszybciej przejrzeć i zdeployować u siebie. Dziękuję!

**Comment by codziennatranskrypcjatvpis** [comment]
> 
Wybacz, że tak późno tutaj zaglądam. Przetestowałem, działa na RPi, wystarczyło dostosować skrypt. Ale deployu nie będzie ze względu na długi czas działania i dość duże zużycie zasobów. Spróbuję zrobić to inaczej: wytnę część zawierającą paski, a następnie usunę wszystkie zduplikowane klatki i dopiero tak zrobiony plik przepuszczę przez opencv (oczywiście dostosuję skrypt).

**Comment by codziennatranskrypcjatvpis** [comment]
Oba rozwiązania są zbyt wolne na RPi. Jakbyś chciał porównać, to moja wersja jest w forku.

**Comment by KMoszczyc** [comment]
Hmm, jak będę miał chwilkę to zobaczę czy da sie to zoptymalizować. Masz jak podejrzeć logi i looknąć co i ile czasu zajmuje?

edit. Patrzyłem na samo wczytywanie mp4 z imutils - FileVideoStream, ale jest wolniejsze od zwykłego OpenCV.
Porobiłem też porównanie ile czasu zajmuje przejście przez wszystkie klatki z mp4 bez przetwarzania vs z przetwarzaniem i wyszło ~ 3.5s vs 5.6s dla 10 minutowego filmiku (z wyłączonym wyświetlaniem klatek). Więc różnica niewielka, ale pewnie na RPi tesseract robi większy narzut. Hm jak już wycinasz te paski to możesz też spróbować zresizować je do mniejszego paska.

**Comment by codziennatranskrypcjatvpis** [comment]
Najnowsza wersja twojego skryptu: 14-20 fps.
Mój pomysł z FFMPEGiem, wycinaniem, zmniejszaniem rozmiaru i zmniejszanie framerate'u: odpada, prędkość ok. 1.1x, czyli za wolno.

**Comment by KMoszczyc** [comment]
Hejka,
pushnąłem nową wersje ze skipowaniem klatek - `tvp_headlines_mp4_skip()`, ustawiłem `FRAMES_TO_SKIP=75` - czyli skipuje 75 klatek -> ocr -> repeat, przy 100 już zaczyna gubić nagłówki.

Dodałem resiza: `resized_img = cv2.resize(binary_img, (0, 0), fx=resize_scale, fy=resize_scale, interpolation=cv2.INTER_AREA)`

Probowałem się pobawić configiem tesseracta by go troche przyśpieszyć, ale bez wielkich sukcesów. 
`headline = pytesseract.image_to_string(resized_img, config=r'--psm 7', lang='pol')`
Tu jest fajny link redditowy:
https://www.reddit.com/r/raspberry_pi/comments/v8w1pf/has_anyone_used_a_rpi4_for_ocr_tesseract_how_did/

U mnie prędkość przetwarzania skoczyła skoczyła do 3.3s na cały filmik. Dodałem też logi tego ile tesseract zajął czasu vs całkowite przetwarzanie. Daj znać co na RPi więcej czasu zajmuje, samo zczytywanie klatek z cap.read() czy tesseract.

**Comment by codziennatranskrypcjatvpis** [comment]
Logi: https://gist.github.com/codziennatranskrypcjatvpis/a2592d42b333fef7bc81284cabdf8410

**Comment by KMoszczyc** [comment]
Hmm z tego co widze to sam OCR Tesseractem zajął stosunkowo mało - 1.7s x 9 ~ 15s, czyli tyle co nic. 92850 *0.028s ~ 2600s - tyle gdzieś zajęła funkcja `extract_headline()` czyli tak z 1/3 całego czasu. Całkowity czas - 6879.10s - kurde serio aż 2h..., czyli 2/3 czasu zajęło samo odczytanie klatek. Też warto wspomnieć, że ja to testowałem na filmiku 640x360 a u ciebie jest 1920x1080, to na pewno ma znaczny wpływ na czas przetwarzania. Więc spróbuj zresizować mp4 ffmpegiem (czy czym tam chcesz) do tego formatu i daj znać jak wtedy pójdzie. + Ten mój plik miał 25fps a twój ma 50, więc możesz zwiększyć parametr FRAMES_TO_SKIP do 150, albo użyć poprzedniej wersji bez skipowania klatek.

**Comment by emilekm2142** [comment]
@KMoszczyc Udało ci się zrobić jakiś zbiór pasków?

**Comment by KMoszczyc** [comment]
@emilekm2142 Nope, ale jest repo na githubie gdzie jest troche starych pasków: https://github.com/Aleshkev/paski-tvp
